### PR TITLE
graphqlbackend: Gracefully handle some code host failures for AffiliatedRepos

### PR DESCRIPTION
**Pull request by ryanslade**
This change allows us to gracefully handle failures in only some of the code
hosts when listing affiliated repos.

The resolver type and file were also renamed to make them more consistent.

Closes: https://github.com/sourcegraph/sourcegraph/issues/20934

Note that the final test is failing which I'll look into tomorrow. It's most likely something to do with
error formatting.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/affiliated_repositories_connection.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/inconshreveable/log15"
+
+	"github.com/sourcegraph/sourcegraph/internal/goroutine"
+
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/backend"
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/affiliated_repositories_connection.go:15]:
```suggestion
	"github.com/pkg/errors"
	"github.com/inconshreveable/log15"

	"github.com/sourcegraph/sourcegraph/internal/goroutine"
	"github.com/sourcegraph/sourcegraph/cmd/frontend/backend"
```

File: cmd/frontend/graphqlbackend/affiliated_repositories_connection.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+			}
+		} else {
+			svc, err := database.ExternalServices(a.db).GetByID(ctx, a.codeHost)
+			if err != nil {
+				a.err = err
+				return
+			}
+			// 🚨 SECURITY: if the user doesn't own this service, check they're site admin
+			if svc.NamespaceUserID != a.userID {
+				if err := backend.CheckUserIsSiteAdmin(ctx, a.userID); err != nil {
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/affiliated_repositories_connection.go:60]: I thought we don't want site admins to see others' config for the Cloud?

**Review by unknwon — CHANGES_REQUESTED** [review]

**Review by unknwon — COMMENTED** [review]
Oh no, hit the wrong button, wanted to just make few comments ...

File: cmd/frontend/graphqlbackend/affiliated_repositories_connection.go (hunk 1 above)
Comment (ryanslade) [reply cmd/frontend/graphqlbackend/affiliated_repositories_connection.go:15]: Ugh, `goimports` never quite gets it right

File: cmd/frontend/graphqlbackend/affiliated_repositories_connection.go (hunk 2 above)
Comment (ryanslade) [reply cmd/frontend/graphqlbackend/affiliated_repositories_connection.go:60]: Yeah, good point. I'll remove this check

**Comment by ryanslade** [comment]
@unknwon The tests were failing because the `RunTests` function used from `graph-gophers` was doing a DeepEqual on internal error state which isn't really under our control. I copied the code out of that package and modified it slightly. 

In a follow-up PR I could switch all our existing uses of `RunTests` to use the modified version which also uses `cmp.Diff` making it a lot easier to spot what's gone wrong. 

WDYT?

File: cmd/frontend/graphqlbackend/testing.go [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strconv"
 	"sync"
 	"testing"
 
+	"github.com/google/go-cmp/cmp/cmpopts"
+
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/testing.go:14]:
```suggestion
```

**Review by unknwon — APPROVED** [review]

**Comment by unknwon** [comment]
> @unknwon The tests were failing because the `RunTests` function used from `graph-gophers` was doing a DeepEqual on internal error state which isn't really under our control. I copied the code out of that package and modified it slightly.
> 
> In a follow-up PR I could switch all our existing uses of `RunTests` to use the modified version which also uses `cmp.Diff` making it a lot easier to spot what's gone wrong.
> 
> WDYT?

If we can "fix" the "RunTests" then it would be great! Previously, I've been just calling the resolver method directly for error cases, and only use "RunTests" for the happy path:

https://github.com/sourcegraph/sourcegraph/blob/8adbbd51aea37c494ee5650c01293a04873be652/cmd/frontend/graphqlbackend/external_services_test.go#L51-L57

**Pull request merged**